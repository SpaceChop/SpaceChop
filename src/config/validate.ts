import console from '../lib/console';
import validate from '../types/validate';

let valid = true;

export default (config, throwError = false) => {
  try {
    validate(config);
    if (!valid) {
      valid = true;
      console.info('Config is valid again');
    }
    return true;
  } catch (err) {
    valid = false;
    console.error('\nThere is an error in /config.yml:');
    console.error('--------------------------------');
    console.error(err.message);
    console.error('--------------------------------');
    if (throwError) {
      throw new Error(err);
    }
  }
  return false;
};
