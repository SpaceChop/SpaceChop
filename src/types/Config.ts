import * as t from 'runtypes';
import validateParams from '../lib/validateParams';
import Paths from './Paths';
import Presets from './Presets';
import Sources from './Sources';
import Storage from './Storage';

const Config = t.Record({
  paths: Paths,
  sources: Sources,
  presets: Presets,
}).And(t.Partial({
  storage: Storage,
  disableChunkedEncoding: t.Boolean,
})).withConstraint(validateParams);

export type Config = t.Static<typeof Config>;
export default Config;
