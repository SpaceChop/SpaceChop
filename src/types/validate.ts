import Config from './Config';

export default (config) => {
  try {
    Config.check(config);
  } catch (err) {
    throw new Error(`${err.key ? `${err.key}: ` : ''}${err.message}`);
  }
};
