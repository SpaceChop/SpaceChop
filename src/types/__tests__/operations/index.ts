import validate from '../..';
import { Config } from '../../Config';

export const defaultConfig: Config = {
  paths: ['/:preset/:image'],
  sources: [{ http: { root: 'http://localhost/:image' }}],
  presets: {},
};

describe('should validate operations', () => {
  it('no double operations in step', () => {
    const config = {
      paths: ['/:preset/:image'],
      sources: [{ http: { root: 'http://localhost' } }],
      presets: {
        t_200: {
          steps: [{
            $crop: {
              width: 100,
              height: 100,
            },
            $resize: {
              width: 100,
              height: 100,
            },
          }],
        },
      },
    };
    expect(() => validate(config)).toThrowErrorMatchingSnapshot();
  });
});
