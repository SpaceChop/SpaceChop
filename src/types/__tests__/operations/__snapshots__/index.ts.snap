// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should validate operations no double operations in step 1`] = `"presets.t_200.steps.[0]: please only one operation per step"`;
